### PR TITLE
feat: enable full hmr support including reload

### PR DESCRIPTION
**Pull request by jonaskuske**
In the first HMR implementation, I noticed that reloading certain components broke with errors like `this.load() is not defined` and just disabled hot component reloads all together by injecting this code that overrides the reload method:

```js
__VUE_HMR_RUNTIME__.reload = () => import.meta.hot.invalidate()
```

I've since investigated the reasons for the errors and made a proper fix/workaround.

### Re-apply Kirby-specific component modifications

The first problem is that Kirby modifies some component definitions before registering them with Vue. It [adds a mixin to section components](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/public/js/plugins.js#L41-L46), [removes the render function if a template is present](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/src/config/plugins.js#L44-L46) so that the template takes precedence and [allows for the `extends` key to reference Kirby built-ins by name](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/src/config/plugins.js#L27-L38), resolving it to the the respective component definition before registering the component. This is also [used used for `block`-type components internally](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/public/js/plugins.js#L19-L22).

When a component is hot reloaded, Vue receives a new copy of the component definition as it is exported from the respective `.vue` file, without Kirby's modifications applied. That is why after a reload, the `.load()` method in a section is suddenly undefined, as it was provided through a mixin that now isn't part of the component definition anymore.

To fix that, we can apply the same changes that Kirby does in [`panel/public/js/plugins.js`](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/public/js/plugins.js) and [`/panel/src/config/plugins.js`](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/src/config/plugins.js) to the incoming new component definition. This is what the new `$_applyKirbyModifications` function does. It is injected into the `reload` method using a RegExp. Since the HMR runtime code rarely changes, this should be stable, but to be sure I've pinned the versions of `@vitejs/plugin-vue2` and `@vitejs/plugin-vue2-jsx` (now included so we can control the version) so they only update if we want them to.

### Let the runtime apply its patch to the correct component definition

The second problem comes from the fact that Kirby not only *modifies* but outright *replaces* some component definitions before registering them. This is the case for [block components](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/public/js/plugins.js#L19-L23) and [section components](https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/panel/public/js/plugins.js#L42-L45), where Kirby doesn't use the component definition object that's supplied by the plugin for registering the component, but instead creates a new object and merges in the one supplied by the plugin using the ...spread operator.

That breaks hot reloads, because the HMR runtime assumes that the component definition object that's exported from a `.vue` file is the one that component instances are based off — it tracks component definitions and their rendered instances in a map like this:

```js
map {
  [someId]: record {
    options: ComponentDefinition, // (the one exported from .vue)
    instances: [VueComponent, VueComponent, ...]
  },
  [someId]: record {
    options: ComponentDefinition, // (the one exported from .vue)
    instances: [VueComponent, VueComponent, ...]
  },
}
```

When a new updated `.vue` module comes in, the runtime patches the definition unter `record.options` to match the definition from the new module, then re-renders all `record.instances` so that the changes show up. However with section and block components, the definition stored under `options` isn't the one the component instances actually use. So while the HMR runtime dutifully patches `record.options` to match the updated component and re-renders everything, those changes never show up because the component definition object created by Kirby, which is the one those instances are *actually* based off, was not modified by the HMR runtime and still has the old template, lifecycle hooks and other properties.

Luckily, all component definitions that are *actually* used by the rendered instances are available via `panel.$vue.$options.components`. So to fix this, we can wrap the `reload` and `rerender` methods in an additional function:
we first check if the updated module belongs to a component added through a panel plugin, and if so check whether the definition registered under `record.options` is the same one as the one registered under `panel.$vue.$options.components[name]`. If it's not, that means `record.options` holds a stale reference and we set it to the actual definition that's in use. After this is done we can proceed to call the original rerender/reload method and perform the HMR update, which now operates on the correct component definition.

With those two changes in place, `kirbyup` is fully compatible with Vue HMR and everything works like it's supposed to. WDYT?

### CODE REVIEW

**Comment by jonaskuske** [comment]
Oh, and while working on this I also noticed that the super handy `--stub` functionality does not work with client-side code, namely `plugin.ts`. Trying to load the stubbed version of the plugin breaks everything. So I disabled stubbing for it and also disabled the `clear` option when the project is stubbed. This way when you run `pnpm dev`, only the Node code is stubbed, the built version of the plugin remains in the dist dir to be imported alongside the stubbed Node code and everything works fine.

Because I've used an env variable to toggle the stubbing, I enabled `shell_emulator` in the npmrc, which tells pnpm to run scripts through [@yarnpkg/shell](https://www.npmjs.com/package/@yarnpkg/shell), enabling cross-browser support for env variables without any additional packages. (like `cross-env`)

**Comment by jonaskuske** [comment]
Also, did a lot of manual testing so should be good, but working on this really underscored that I finally need to add some HMR tests 😅

**Comment by johannschopplich** [comment]
> WDYT?

Hey, Jonas, outstanding additions to kirbyup! 💛 Thank you sou much for your contribution and this amazing code.

I have to admit, I don't get the changes 100 %, since I'm not that deep into Vite's HMR. Thanks for the detailed explanation on what is happening! 

Also, thanks for enabling cross env support! I've read about it in the pnpm docs some time ago.

> Also, did a lot of manual testing so should be good, but working on this really underscored that I finally need to add some HMR tests 😅

I trust your tests. Also, writing comprehensive tests for HMR inside Kirby seems like a complicated task.

File: src/node/plugins/utils.ts [hunk 1]
```diff
@@ -1,2 +1,106 @@
...
+    else if (usedComponentDefs[newDef.extends]) {
+      newDef.extends = usedComponentDefs[newDef.extends].extend({
+        options: newDef,
+        components: { ...usedComponentDefs, ...(newDef.components || {}) },
+      })
+    }
+    else { newDef.extends = null }
+  }
+}
+/** */
```
Comment (johannschopplich) [inline src/node/plugins/utils.ts:105]: Last question: Is this comment intentional or a leftover?

**Comment by johannschopplich** [comment]
I've released a new beta for your latest changes: [`v2.1.0-beta.0`](https://github.com/johannschopplich/kirbyup/releases/tag/v2.1.0-beta.0)

**Pull request merged**